src/main/java/com/artipie/asto/ArtipieStorage.java

```diff
@@ -0,0 +1,249 @@
+                final CompletableFuture<Void> term = new CompletableFuture<>();
+                if (status.success()) {
+                    new PublisherAs(
+                        Flowable.fromPublisher(rsbody)
+                            .doOnError(term::completeExceptionally)
+                            .doOnTerminate(() -> term.complete(null))
+                    ).string(StandardCharsets.UTF_8)
+                        .thenApply(s -> promise.complete(ArtipieStorage.parse(s)));
+                } else {
+                    final String msg = String.format(
+                        "Cannot get lists blobs contained in given path [prefix=%s, status=%s]",
+                        prefix,
+                        status
+                    );
+                    promise.completeExceptionally(
+                        new ArtipieIOException(msg)
+                    );
+                }
+                return term;
+            }
+        );
+        return promise;
+    }
+
+    @Override
+    public CompletableFuture<Void> save(final Key key, final Content content) {
+        return this.remote.response(
+            new RequestLine(RqMethod.PUT, ArtipieStorage.uri(key)).toString(),
+            new Headers.From(new ContentLength(content.size().get())),
+            content
+        ).send(
+            (status, rsheaders, rsbody) -> {
+                final CompletableFuture<Void> res;
+                if (status.success()) {
+                    res = CompletableFuture.allOf();
+                } else {
+                    res = new CompletableFuture<>();
+                    res.completeExceptionally(
+                        new ArtipieIOException(
+                            String.format(
+                                "Entry is not created [key=%s, status=%s]",
+                                key,
+                                status
+                            )
+                        )
+                    );
+                }
+                return res;
+            }
+        ).toCompletableFuture();
+    }
+
+    @Override
+    public CompletableFuture<Void> move(final Key source, final Key destination) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletableFuture<? extends Meta> metadata(final Key key) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public CompletableFuture<Content> value(final Key key) {
+        final CompletableFuture<Content> promise = new CompletableFuture<>();
+        this.remote.response(
+            new RequestLine(RqMethod.GET, ArtipieStorage.uri(key)).toString(),
+            Headers.EMPTY,
+            Content.EMPTY
+        ).send(
+            (status, rsheaders, rsbody) -> {
+                final CompletableFuture<Void> term = new CompletableFuture<>();
+                if (status.success()) {
+                    promise.complete(
+                        new Content.From(
```

Comment:
we have `ContentWithSize` class, that accepts publisher and headers, I'd suggest to use it here and where possible in this class